uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-next-leave.component.js

```diff
@@ -0,0 +1,143 @@
+    'shared-settings'];
+
+  function nextLeaveController ($q, $scope, LeaveRequest, OptionGroup,
+  sharedSettings) {
+    var childComponentName = 'leave-widget-next-leave';
+    var vm = this;
+
+    vm.dayTypes = {};
+    vm.balanceDeduction = 0;
+    vm.nextLeaveRequest = null;
+    vm.requestStatus = {};
+
+    vm.$onChanges = $onChanges;
+
+    /**
+     * Initializes the controller by emiting a child is loading event.
+     */
+    (function init () {
+      $scope.$emit('LeaveWidget::childIsLoading', childComponentName);
+    })();
+
+    /**
+     * Implements $onChanges method for AngularJS Components. Waits for bindings
+     * to be ready, and then loads the next leave requests. When it's done, it
+     * emits a child is ready event.
+     */
+    function $onChanges () {
+      if (!bindingsAreReady()) {
+        return;
+      }
+
+      $q.all([
+        loadDayTypes(),
+        loadNextLeaveRequest()
+      ])
+      .then(function () {
+        if (vm.nextLeaveRequest) {
+          makeBalanceChangeAbsolute();
+          storeStatusForNextRequest();
+        }
+      })
+      .finally(function () {
+        $scope.$emit('LeaveWidget::childIsReady', childComponentName);
+      });
+    }
+
+    /**
+     * Returns true when contact id and leave request statuses bindings are
+     * ready.
+     *
+     * @return {Boolean}
+     */
+    function bindingsAreReady () {
+      return vm.contactId && vm.leaveRequestStatuses &&
+        vm.leaveRequestStatuses.length;
+    }
+
+    /**
+     * Returns a list of status ids.
+     *
+     * @return {Array}
+     */
+    function getStatusIds () {
+      return vm.leaveRequestStatuses.map(function (status) {
+        return status.value;
+      });
+    }
+
+    /**
+     * Loads and stores all the possible day types for leave requests indexed
+     * by value.
+     *
+     * @return {Promise}
+     */
+    function loadDayTypes () {
+      return OptionGroup.valuesOf('hrleaveandabsences_leave_request_day_type')
+        .then(function (_dayTypes_) {
+          vm.dayTypes = _.indexBy(_dayTypes_, 'value');
+        });
+    }
+
+    /**
+     * Loads the next leave request for the contact that has been approved or
+     * is under review.
+     *
+     * @return {Promise}
+     */
+    function loadNextLeaveRequest () {
+      var statusIds = getStatusIds();
+      var today = moment().format(sharedSettings.serverDateFormat);
+
+      return LeaveRequest.all({
+        contact_id: vm.contactId,
+        from_date: { '>=': today },
+        request_type: 'leave',
+        status_id: { IN: statusIds },
+        options: { limit: 1, sort: 'from_date DESC' }
+      })
+      .then(function (response) {
+        vm.nextLeaveRequest = response.list[0] || null;
+      });
+    }
+
+    /**
+     * Makes the next leave requst balance chance value absolute.
+     */
+    function makeBalanceChangeAbsolute () {
+      vm.nextLeaveRequest = _.defaults({
```

Comment:
To be honest this is done so other tests don't break by overwriting the mock values. Overwriting the value directly would not normally affect the app otherwise.

Reply:
then there is something wrong with the tests? We can't the tests dictate the implementation